app/src/main/java/com/termux/app/terminal/TermuxActivityRootView.java

```diff
@@ -0,0 +1,227 @@
+ * Also check https://github.com/termux/termux-app/issues/1539.
+ *
+ * This overlap may happen even without `enforce-char-based-input=true` for keyboards with extended layouts
+ * like number row, etc.
+ *
+ * To fix these issues, `activity_termux.xml` has the constant 1sp transparent
+ * `activity_termux_bottom_space_view` View at the bottom. This will appear as a line matching the
+ * activity theme. When {@link TermuxActivity} {@link ViewTreeObserver.OnGlobalLayoutListener} is
+ * called when any of the sub view layouts change,  like keyboard opening/closing keyboard,
+ * extra keys/input view switched, etc, we check if the bottom space view is visible or not.
+ * If its not, then we add a margin to the bottom of the root view, so that the keyboard does not
+ * overlap the extra keys/terminal, since the margin will push up the view. By default the margin
+ * added is equal to the height of the hidden part of extra keys/terminal. For Gboard's case, the
+ * hidden part equals the `header_height`. The updates to margins may cause a jitter in some cases
+ * when the view is redrawn if the margin is incorrect, but logic has been implemented to avoid that.
+ */
+public class TermuxActivityRootView extends LinearLayout implements ViewTreeObserver.OnGlobalLayoutListener {
+
+    public TermuxActivity mActivity;
+    public Integer marginBottom;
+    public Integer lastMarginBottom;
+
+    /** Log root view events. */
+    private boolean ROOT_VIEW_LOGGING_ENABLED = false;
+
+    private static final String LOG_TAG = "TermuxActivityRootView";
+
+    public TermuxActivityRootView(Context context) {
+        super(context);
+
+    }
+
+    public TermuxActivityRootView(Context context, @Nullable AttributeSet attrs) {
+        super(context, attrs);
+    }
+
+    public TermuxActivityRootView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
+        super(context, attrs, defStyleAttr);
+    }
+
+    public void setActivity(TermuxActivity activity) {
+        mActivity = activity;
+    }
+
+    /**
+     * Sets whether root view logging is enabled or not.
+     *
+     * @param value The boolean value that defines the state.
+     */
+    public void setIsRootViewLoggingEnabled(boolean value) {
+        ROOT_VIEW_LOGGING_ENABLED = value;
+    }
+
+    @Override
+    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
+        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
+
+        if (marginBottom != null) {
+            if (ROOT_VIEW_LOGGING_ENABLED)
+                Logger.logVerbose(LOG_TAG, "onMeasure: Setting bottom margin to " + marginBottom);
+            ViewGroup.MarginLayoutParams params = (ViewGroup.MarginLayoutParams) getLayoutParams();
+            params.setMargins(0, 0, 0, marginBottom);
+            setLayoutParams(params);
+            marginBottom = null;
+            requestLayout();
+        }
+    }
+
+    @Override
+    public void onGlobalLayout() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 21. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.